### PR TITLE
HazelcastJsonValue implementation

### PR DESCRIPTION
**Pull request by stalek71**

### CODE REVIEW

**Comment by stalek71** [comment]
verify

File: Hazelcast.Examples/App.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
```
Comment (asimarslan) [inline Hazelcast.Examples/App.cs]: We should update the copyright in a separate PR.

Comment (stalek71) [reply Hazelcast.Examples/App.cs]: It has been sent as this PR:
https://github.com/hazelcast/hazelcast-csharp-client/pull/226

**Comment by asimarslan** [comment]
verify

File: Hazelcast.Examples/App.cs (hunk 1 above)
Comment (asimarslan) [reply Hazelcast.Examples/App.cs]: Please remove this change from this PR

File: Hazelcast.Net.sln [hunk 2]
```diff
@@ -14,8 +14,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		build-docs.bat = build-docs.bat
 		build-nuget.bat = build-nuget.bat
 		build.bat = build.bat
-		shared.project.props = shared.project.props
```
Comment (asimarslan) [inline Hazelcast.Net.sln]: Do we need this change?

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJson.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Hazelcast.IO.Serialization
+{
+    /// <summary>
+    ///  Class for serializing/deserializing Hazelcast JSON types
+    /// </summary>
+    public class HazelcastJson
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJson.cs]: Please update this class according to design doc

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs [hunk 4]
```diff
@@ -349,13 +348,13 @@ protected internal ISerializerAdapter SerializerFor(int typeId)
             if (typeId <= 0)
             {
                 var index = IndexForDefaultType(typeId);
-                if (index < ConstantSerializersSize)
-                {
+                if (index < ConstantSerializersSize && index != -SerializationConstants.DefaultTypeSerializable)
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: missing curly bracket

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs [hunk 5]
```diff
@@ -682,11 +684,11 @@ private ISerializerAdapter SerializerFor(object obj)
             if (serializer == null)
             {
                 if (_isActive)
-                {
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: please do not format or remove curly bracket from unrelated code like `SerializerFor` methods

File: Hazelcast.Test/Hazelcast.Client.Test/ClientMapJsonTest.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+using Hazelcast.Config;
+using Hazelcast.Core;
+using Hazelcast.IO.Serialization;
+using NUnit.Framework;
+
+namespace Hazelcast.Client.Test
+{
+    [TestFixture]
+    public class ClientMapJsonTest : SingleMemberBaseTest
```
Comment (asimarslan) [inline Hazelcast.Test/Hazelcast.Client.Test/ClientMapJsonTest.cs]: Missing category

File: Hazelcast.Examples/App.cs (hunk 1 above)
Comment (stalek71) [reply Hazelcast.Examples/App.cs]: Changed back to 2018

File: Hazelcast.Net.sln (hunk 2 above)
Comment (stalek71) [reply Hazelcast.Net.sln]: No, we don't need this change. It was some change done by VS. I will revert it but the question is if it updates solution config it preserves this order or introduces this change (or similar) again...

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs (hunk 4 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: fixed

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs (hunk 5 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: added missing brackets

File: Hazelcast.Test/Hazelcast.Client.Test/ClientMapJsonTest.cs (hunk 6 above)
Comment (stalek71) [reply Hazelcast.Test/Hazelcast.Client.Test/ClientMapJsonTest.cs]: Added ``[Category("3.11")]``

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationConstants.cs [hunk 7]
```diff
@@ -48,7 +48,10 @@ internal sealed class SerializationConstants
 
         public const int ConstantSerializersLength = 28;
 
-        public const int DefaultTypeSerializable = -110; //C# serializable
+        public const int DefaultTypeSerializable = -110; // C# serializable
+        public const int DefaultTypeJsonString = -130;
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/SerializationConstants.cs]: On java side we used the name `JAVASCRIPT_JSON_SERIALIZATION_TYPE`. Can you match the same name?

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs [hunk 8]
```diff
@@ -349,13 +348,15 @@ protected internal ISerializerAdapter SerializerFor(int typeId)
             if (typeId <= 0)
             {
                 var index = IndexForDefaultType(typeId);
-                if (index < ConstantSerializersSize)
+                if (index < ConstantSerializersSize && index != -SerializationConstants.DefaultTypeSerializable)
                 {
                     return _constantTypeIds[index];
                 }
             }
+
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: We don't use empty lines like this. I can see a couple of them below too. 
Please do not reformat the code unless we change the formatting rules.

File: Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hazelcast.Client.Test.Serialization
+{
+    [TestFixture]
+    public class JsonValueSerializationTest
```
Comment (asimarslan) [inline Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs]: Missing category

File: README.md [hunk 10]
```diff
@@ -840,7 +843,40 @@ clientConfig.GetSerializationConfig()
...
+
+```c#
+    var config = new ClientConfig();
+    var client = HazelcastClient.NewHazelcastClient(config);
+    var map = client.GetMap<string, HazelcastJsonValue>("map");
+```
+
+We constructed a map in the cluster which has `string` as the key and `HazelcastJsonValue` as the value. `HazelcastJsonValue` is a simple wrapper and identifier for the JSON formatted strings. You can get the JSON string from the `HazelcastJsonValue` object by using the `JsonString` property. 
+
+You can construct a `HazelcastJsonValue` using one of the constructors. All constructors accept the JSON formatted string as the parameter. No JSON parsing is performed but it is your responsibility to provide correctly formatted JSON strings. The client will not validate the string, and it will send it to the cluster as it is. If you submit incorrectly formatted JSON strings and, later, if you query those objects, it is highly possible that you will get formatting errors since the server will fail to deserialize or find the query fields.
```
Comment (asimarslan) [inline README.md]: We don't provide constructors, there is only one static Factory method. This is different from C++

File: Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs (hunk 9 above)
Comment (asimarslan) [reply Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs]: `HazelcastJsonValueSerializationTest` is a better name I guess.

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationConstants.cs (hunk 7 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationConstants.cs]: The name has been changed to `JavascriptJSONSerializationType`

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs (hunk 8 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: Removed

File: README.md (hunk 10 above)
Comment (stalek71) [reply README.md]: Fixed. Strange. I was sure I changed the fragment about constructors already...

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJson.cs (hunk 3 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJson.cs]: Refactored

File: Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs (hunk 9 above)
Comment (stalek71) [reply Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs]: Renamed

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs (hunk 8 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: Can you remove other spaces you added?

Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: Restored to original state

**Comment by stalek71** [comment]
verify

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        protected bool Equals(HazelcastJsonValue other)
+        {
+            return string.Equals(_jsonString, other._jsonString);
+        }
+
+        public static HazelcastJsonValue FromString(string jsonString)
+        {
+            return new HazelcastJsonValue(
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: It's better to keep the validation on a separate line.

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                ValidationUtil.CheckNotNull(jsonString, ValidationUtil.NULL_VALUE_IS_NOT_ALLOWED));
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]:
```
 ValidationUtil.NULL_JSON_STRING_IS_NOT_ALLOWED = "JSON string cannot be null."
...
ValidationUtil.CheckNotNull(jsonString, ValidationUtil.NULL_JSON_STRING_IS_NOT_ALLOWED);
return new HazelcastJson(jsonString);
```

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+namespace Hazelcast.IO.Serialization
+{
+    /// <summary>
+    ///  Class for serializing/deserializing Hazelcast JSON types
+    /// </summary>
+    public class HazelcastJsonValue
+    {
+        private string _jsonString;
+
+        internal HazelcastJsonValue(string json)
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: this should be private to force static method usage that does a null validation from everywhere

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Hazelcast.Util;
+using System;
+
+namespace Hazelcast.IO.Serialization
+{
+    /// <summary>
+    ///  Class for serializing/deserializing Hazelcast JSON types
```
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: API doc need to be improved. Please use the doc on Java API and document other methods:
`ToString` `FromString`

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs (hunk 4 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: index check would be better with:
`index < ConstantSerializersSize && index != IndexForDefaultType(SerializationConstants.DefaultTypeSerializable)`

File: Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+            var jsonData = _serializationService.ToData(jsonValue);
+            var jsonDeserialized = _serializationService.ToObject<HazelcastJsonValue>(jsonData);
+
+            Assert.AreEqual(jsonValue, jsonDeserialized);
+        }
+
+        [Test]
+        public void FromStringShouldThrowExceptionOnNullArgument()
+        {
+            Exception ex = null;
```
Comment (asimarslan) [inline Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs]: NUnit Assert to check exceptions:
```
 Assert.Throws<NullReferenceException>(() => HazelcastJsonValue.FromString(null));
```

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs (hunk 11 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: This `ValidationUtil.CheckNotNull()` sends you back the result if it's valid so it's better to use it in this dedicated form. On separate line it looks we use just `side effect` of this function

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs (hunk 13 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: We use this constructor in default serializers class also (so the same lib) -  DefaultSerializers.cs:53

File: Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs (hunk 15 above)
Comment (stalek71) [reply Hazelcast.Test/Hazelcast.Client.Test.Serialization/JsonValueSerializationTest.cs]: Fixed

File: Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs (hunk 4 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/SerializationService.cs]: Changed

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs (hunk 14 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: Updated

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs (hunk 13 above)
Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: Changed to factory method

File: Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs (hunk 11 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: Yes we'll use the side effect you don't need any return value. if it is null then it will throw an exception.
You can check the other usages of this method.

Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: Ok. I will send such change in a moment

Comment (stalek71) [reply Hazelcast.Net/Hazelcast.IO.Serialization/HazelcastJsonValue.cs]: Changed
